clients/src/main/java/org/apache/kafka/clients/admin/AdministrativeClient.java

```
@@ -0,0 +1,628 @@
+        }
+    }
+
+    /**
+     * List topics in the cluster.
+     *
+     * @param flags                     The flags to use on the request.
+     * @return                          A map from topic names to TopicInfo objects or errors.
+     *                                  An error is returned when we are not authorized to describe the given topic.
+     * @throws InterruptedException     If the thread receives an InterruptedException before finishing.
+     * @throws TimeoutException         If the thread times out before receiving cluster metadata.
+     */
+    public Map<String, Try<TopicInfo>> listTopics(EnumSet<ListTopicsFlags> flags) throws InterruptedException {
+        final boolean includeInternal = flags.contains(ListTopicsFlags.INCLUDE_INTERNAL);
+        return waitForResultFromController("listTopics",
+            new FutureProvider<RequestFuture<Map<String, Try<TopicInfo>>>>() {
+                @Override
+                public RequestFuture<Map<String, Try<TopicInfo>>> provide(Node controller) {
+                    final MetadataRequest.Builder builder = MetadataRequest.Builder.allTopics();
+                    return consumerClient.send(controller, builder).
+                        compose(new RequestFutureAdapter<ClientResponse, Map<String, Try<TopicInfo>>>() {
+                            @Override
+                            public void onSuccess(ClientResponse response, RequestFuture<Map<String, Try<TopicInfo>>> future) {
+                                MetadataResponse mtr = (MetadataResponse) response.responseBody();
+                                HashMap<String, Try<TopicInfo>> results = new HashMap<>();
+                                for (MetadataResponse.TopicMetadata topic : mtr.topicMetadata()) {
+                                    if (topic.isInternal() && !includeInternal)
+                                        continue;
+                                    if (topic.error() != Errors.NONE) {
```

is it possible that this masks errors in `partitionMetadata`? e.g. if we have a topic level error and an error on a specific partition.

---

Hmm, I'm not sure I follow.  All the errors in the RPC response are faithfully reproduced in the return values... what else can we do to improve it?

---

you have a `continue` in this block. my question is related to if you end up with a `TopicMetadata` that looks like:
```
{
  topic: 'mytopic',
  error: Error.SOMETHING_BAD,
  partitionMetadata: [
    {
      partition: 0
      error: SOMETHING_ELSE_THAT_IS_BAD
    } 
  ]
}
```

also, in writing this up, i realize that in line 487 you are checking `topic.error()` instead of `partition.error()`